### ISSUE TITLE
child process exited abnormally06304

### ISSUE THREAD

**Issue by melmacjim** [ianka/xfs_undelete#1]
I was trying to use `xfs_undelete` to see if it could uncover directories and files that we either deleted, overwritten, or the drive might have had its MBR overwritten.

I haven't been able to get past the following:
```
# ./xfs_undelete -o /scr2/recovered-data-l211 /dev/sdb1
--
child process exited abnormally06304  (  0.0%)
while executing
"exec -ignorestderr -- dd 2>/dev/null if=$fs of=$of bs=$blocksize skip=$skip seek=$loffset count=$count"
("for" body line 43)
invoked from within
"for {set block [dict get $::parameters s]} {$block<$dblocks} {incr block} {
## Log each visited block.
puts -nonewline stderr [format $m1format $blo..."
(file "./xfs_undelete" line 59)
```
From your code: 
```
...
 41 foreach line [split $config \n] {
 42         lassign $line key dummy value
 43         if {$key in {blocksize inodesize agblocks agblklog dblocks}} {
 44                 set $key $value
 45         }
 46 }
...
 58 ## Run through whole filesystem.
 59 for {set block [dict get $::parameters s]} {$block<$dblocks} {incr block} {
...
```

Some background:
I'm not sure what happened with the data, but the user says there were multiple directories with multiple files and now they're all gone. The suspect commands in the bash history are `rm -rf blah` which is followed by a `fdisk /dev/THE-DRIVE`. There weren't any timestamps in the bash history, so those `rm` and `fdisk` commands could have been from months ago during provisioning.
 

System details:
OS: CentOS7
HDD: 8TB Western Digital Red
Partition Details: 
```
# fdisk -l /dev/sdb
...
Disk /dev/sdb: 8001.6 GB, 8001563222016 bytes, 15628053168 sectors
Units = sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 4096 bytes
I/O size (minimum/optimal): 4096 bytes / 4096 bytes
Disk label type: gpt
Disk identifier: E7F55CB0-5C0B-43E3-9978-438D59CAFEDB


#         Start          End    Size  Type            Name
 1         2048  15628052479    7.3T  Microsoft basic primary
```

When mounted:
```
# mount /dev/sdb1 /TEST-DATA-RECOVERY

# mount -l | grep sdb
/dev/sdb1 on /TEST-DATA-RECOVERY type xfs (rw,relatime,attr2,inode64,noquota) [/label1]

# ls -la /TEST-DATA-RECOVERY/
total 4
drwxrws---   2 root psgvb    6 Jul 25 14:11 .
dr-xr-xr-x. 26 root root  4096 Jul 26 16:14 ..

# df -lH | grep sdb
/dev/sdb1       8.0T   35M  8.0T   1% /TEST-DATA-RECOVERY
```

Any thoughts?

Thanks.

**Comment by ianka** [comment]
The script complains about ```dd``` terminating unexpectedly. You could change the ```2>/dev/null``` in line 101 into ```2>mylogfile``` to get some information on the problem.

Also check the kernel log. Maybe this drive has a hardware problem and the driver simply aborts ```dd```.

Another problem I see is that inode64 mount option. This might be incompatible with the inode/extent format the script understands and/or with ```dd```. I have to look that up.

**Comment by ianka** [comment]
I checked on the inode64 option. It shouldn't be a problem. The filesystem I recovered my data from with that script was mounted with inode64, too. I used dd-8.31 and recovered the data on a 64 bit system.

**Comment by mididix** [comment]
Had the same issue right now on my FS:

```
[root@fedora-dg xfs_undelete]# ./xfs_undelete /dev/mapper/fedora_linux-root
Checking block     383/5890048  (  0.0%)
Recovered deleted inode 3065.
Checking block    8270/5890048  (  0.1%)
Recovered deleted inode 66162.
child process exited abnormally (  0.2%)
    while executing
"exec -ignorestderr -- dd 2>/tmp/xfsundelete if=$fs of=$of bs=$blocksize skip=$skip seek=$loffset count=$count"
    ("for" body line 43)
    invoked from within
"for {set block [dict get $::parameters s]} {$block<$dblocks} {incr block} {
	## Log each visited block.
	puts -nonewline stderr [format $m1format $blo..."
    (file "./xfs_undelete" line 59)
```

Here is the error from dd:
```
dd: offset too large: cannot truncate to a length of seek=15365922889924608 (4096-byte) blocks
```

Any idea?

**Comment by ianka** [comment]
Some inodes contain extents with invalid block numbers as this. xfs_undelete has a simple sanity check at [line98](https://github.com/ianka/xfs_undelete/blob/master/xfs_undelete#L98) but only for the ```skip=``` parameter to ```dd```. Could you try adding another line right below line 98

```if {($loffset+$count)>=$dblocks} continue```

This isn't entirely correct as such huge files may exist as sparse files, but it could fix it for you.

**Comment by ianka** [comment]
I added a fix for those problems with dd. It just ignores those extents now and reports on stderr. Please update to v1.3 and try again.

**Comment by mididix** [comment]
Yup, it worked :)
With a bunch of "cannot seek: Value too large for defined data type", but it got through to the end.

Thanks.

**Comment by ianka** [comment]
Those files most likely aren't recoverable anyways.

As the original posters' problem should be fixed the same way, I close this issue.

**Comment by ianka** [comment]
Fixed with version 1.3.
